mathics/builtin/numbers/calculus.py

```diff
@@ -1631,3 +1947,81 @@ def apply_makeboxes(self, x, x0, data, nmin, nmax, den, form, evaluation):
         # expansion = [ex.format(form) for ex in expansion]
         expansion = Expression(SymbolPlus, *expansion)
         return expansion.format(evaluation, form)
+
+
+# Auxiliary routines. Maybe should be moved to another module.
+
+
+def get_accuracy_and_prec(opts: dict, evaluation: "Evaluation"):
+    """
+    Looks at an opts dictionary and tries to determine the numeric values of
+    Accuracy and Precision goals. If not available, returns None.
+    """
+    acc_goal = opts.get("System`AccuracyGoal", None)
+    if acc_goal:
+        acc_goal = apply_N(acc_goal, evaluation)
+        if acc_goal is SymbolAutomatic:
+            acc_goal = Real(12.0)
+        elif acc_goal is SymbolInfinity:
+            acc_goal = None
+        elif not isinstance(acc_goal, Number):
+            acc_goal = None
+
+    prec_goal = opts.get("System`PrecisionGoal", None)
+    if prec_goal:
+        prec_goal = apply_N(prec_goal, evaluation)
+        if prec_goal is SymbolAutomatic:
+            prec_goal = Real(12.0)
+        elif prec_goal is SymbolInfinity:
+            prec_goal = None
+        elif not isinstance(prec_goal, Number):
+            prec_goal = None
+    return acc_goal, prec_goal
+
+
+def is_zero(val, acc_goal, prec_goal, evaluation):
+    """
+    Check if val is zero upto the precision and accuracy goals
+    """
+    if not isinstance(val, Number):
+        val = apply_N(val, evaluation)
+    if not val.is_numeric():
+        return False
+    if val.is_zero:
+        return True
+    if acc_goal:
```

Comment:
We've seen this convoluted kind of nested if's before. Can't the two pieces of code be used in one function?